XenAdmin/Diagnostics/Checks/UpgradeRequiresEUA.cs

```diff
@@ -0,0 +1,75 @@
+﻿/* Copyright (c) Cloud Software Group, Inc. 
+ * 
+ * Redistribution and use in source and binary forms, 
+ * with or without modification, are permitted provided 
+ * that the following conditions are met: 
+ * 
+ * *   Redistributions of source code must retain the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer. 
+ * *   Redistributions in binary form must reproduce the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer in the documentation and/or other 
+ *     materials provided with the distribution. 
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND 
+ * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using XenAdmin.Core;
+using XenAPI;
+using XenAdmin.Diagnostics.Problems;
+using XenAdmin.Diagnostics.Problems.UtilityProblem;
+
+namespace XenAdmin.Diagnostics.Checks
+{
+    internal class UpgradeRequiresEua : UpgradeCheck
+    {
+        private readonly Uri _targetUri;
+        public UpgradeRequiresEua(List<Host> hosts, Uri targetUri)
+            : base(hosts)
+        {
+            _targetUri = targetUri;
+        }
+
+        public override bool CanRun() => Hosts.Any(Helpers.Post82X);

```

Comment:
> This means the check is dependent on installing the RPU hotfix first (see for example SafeToUpgradeCheck).

For this, I was planning to add the relevant bits once we had an RPU (so within the RPU PR)

> If there is an error while retrieving the EUA though, I don't think we shoud proceed suddenly.

Unfortunately, the `getEUA` function returns `{ "eua" : null }` both in case of error and in case there's no EUA at all. We can circumvent this because we know that if the RPU is for Nile or greater the user needs to accept an EUA, but that is logic we'll need to add to XenCenter as opposed to relying on the `getEUA` function.

Not ideal in my opinion, since when/if things change we'll need to update both (e.g.: no EUA for a super special release? One never knows what's going to happen with these things 😅).

So to recap if XenCenter cannot fetch an EUA but we except one, we can add special wording instead of showing nothing. I was thinking something on the lines of `XenCenter could not fetch an End User Agreement (EUA), but expected one. Please go to https://example.com and check if your installation requires accepting an EUA. Selecting "I accept" means you have read and accept the terms of the EUA, if there is one.`

Reply:
I just noticed we already do something like this in the `AcceptEuaDialog`, where I show an error if the host is POST82X and we can't fetch the EUA

Reply:
Yes, something like the text you suggest above is exactly what I had in mind, I didn't imply adding the EUA within XC.
The error currently shown on AcceptEuaDialog does not cover us.